### PR TITLE
Api to clear all answers of questionnaire response.

### LINKED ISSUE #2221: Clear all action on Questionnaire
**Is your feature request related to a problem? Please describe.**
We'd like to reset all the answered fields in the questionnaires via a `Clear all` action.

**Describe the solution you'd like**

The "Clear all" function will remove all the answer items for the "current" QuestionnaireResponse when the questionnaire is being used actively.

**Describe alternatives you've considered**
None

**Additional context**
In order to filter the content of our registers (list of resources) we opted to re-use the UI rendered via Questionnaires instead of implementing custom UI to capture the content for the filter. As part of the requirement, a "Clear all" action is required to reset all the data filter fields".  

See the screenshot below, the Questionnaire is the second screen in the pic.

image (1)

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2221 

**Description**
Provide api to clear all answers present in the questionnaire response.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -421,6 +421,27 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
+   * Recursively clears the answer components of a questionnaire response item by iterating through
+   * each child item.
+   *
+   * @param item The item component of the questionnaire response to clear.
+   */
+  private fun clearItemAnswerComponent(item: QuestionnaireResponseItemComponent) {
+    for (child in item.item) {
+      clearItemAnswerComponent(child)
+    }
+    item.answer = listOf()
```
Comment (MJ1998) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:442]: You can use `QuestionnaireResponse.allItems` defined in `MoreQuestionnaireResponse`
```suggestion
    questionnaireResponse.allItems().forEach { it.answer = emptyList() }
    modificationCount.update { it + 1 }
```

**Review by MJ1998 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 1 above)
Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: done.

**Review by MJ1998 — COMMENTED** [review]
What about draftAnswerMap, modifiedQuestionnaireResponseItemSet, responseItemToAnswersMapForDisabledQuestionnaireItem ,etc ?
Do they need to be cleared as well ?

**Comment by santosh-pingle** [comment]
> What about draftAnswerMap, modifiedQuestionnaireResponseItemSet, responseItemToAnswersMapForDisabledQuestionnaireItem ,etc ? Do they need to be cleared as well ?

I am trying to understand why it is necessary to clear these data structures individually. Is there any use case that requires attention? Wouldn't calling this function modificationCount.update { it + 1 } be sufficient?

**Comment by MJ1998** [comment]
I am suspecting that draftAnswers can still show up. 

Notice the `answersChangedCallback`. Whenever any answer changes we updated draftAnswer, modifiedQuestionnaireResponseItemSet and then do `modificationCount.update { it + 1 }`.
So I think draftAnswer and modifiedQuestionnaireResponseItemSet definitely needs to be cleared otherwise the obsolete values will still show. Try a testcase where you can have draftAnswers, eg, edittext.

Again, the cached answers in `responseItemToAnswersMapForDisabledQuestionnaireItem` should  also be cleared even though I don't think it will effect anything but no point keeping obsolete answers.

**Comment by santosh-pingle** [comment]
> I am suspecting that draftAnswers can still show up.
> 
> Notice the `answersChangedCallback`. Whenever any answer changes we updated draftAnswer, modifiedQuestionnaireResponseItemSet and then do `modificationCount.update { it + 1 }`. So I think draftAnswer and modifiedQuestionnaireResponseItemSet definitely needs to be cleared otherwise the obsolete values will still show. Try a testcase where you can have draftAnswers, eg, edittext.
> 
> Again, the cached answers in `responseItemToAnswersMapForDisabledQuestionnaireItem` should also be cleared even though I don't think it will effect anything but no point keeping obsolete answers.

Ohhh!, thats correct, it requires to clear draftAnswer or else it keep showing partial answer in the textfields, e.g partial date answer in the datePicker. Basically, clears dataStructures which get modified by answerChangeCallback.

**Pull request merged**